src/lib/OS/SyscallHandler.cc

```diff
@@ -1,15 +1,711 @@
+      int fd = info.registerArguments[4].get<int>();
+      off_t offset = info.registerArguments[5].get<off_t>();
+
+      uint64_t result = mmap(addr, length, prot, flags, fd, offset);
+      if (result <= 0) {
+        stateChange = {
+            ChangeType::REPLACEMENT, {info.ret}, {static_cast<int64_t>(-1)}};
+      } else {
+        stateChange = {ChangeType::REPLACEMENT, {info.ret}, {result}};
+      }
+      break;
+    }
+    case 226: {  // mprotect
+      // mprotect is not supported
+      // always return zero to indicate success
+      stateChange = {ChangeType::REPLACEMENT, {info.ret}, {0ull}};
+      break;
+    }
+    case 235: {  // mbind
+      // mbind is not supported due to all binaries being single threaded.
+      // Always return zero to indicate success
+      stateChange = {ChangeType::REPLACEMENT, {info.ret}, {0ull}};
+      break;
+    }
+    case 261: {  // prlimit64
+      // TODO: Functionality temporarily omitted as it is unused within
+      // workloads regions of interest and not required for their simulation
+      stateChange = {ChangeType::REPLACEMENT, {info.ret}, {0ull}};
+      break;
+    }
+    case 278: {  // getrandom
+      // TODO: support flags argument
+
+      // seed random numbers
+      srand(clock());
+
+      // Write <buflen> random bytes to buf
+      uint64_t bufPtr = info.registerArguments[0].get<uint64_t>();
+      size_t buflen = info.registerArguments[1].get<size_t>();
+
+      char buf[buflen];
+      for (size_t i = 0; i < buflen; i++) {
+        buf[i] = (uint8_t)rand();
+      }
+
+      stateChange = {ChangeType::REPLACEMENT, {info.ret}, {(uint64_t)buflen}};
+
+      stateChange.memoryAddresses.push_back({bufPtr, (uint8_t)buflen});
+      stateChange.memoryAddressValues.push_back(RegisterValue(buf, buflen));
+
+      break;
+    }
+    case 293: {  // rseq
+      stateChange = {ChangeType::REPLACEMENT, {info.ret}, {0ull}};
+      break;
+    }
+
+    default:
+      break;
+  };
+
+  concludeSyscall(stateChange);
+}
+
+void SyscallHandler::readStringThen(char* buffer, uint64_t address,
+                                    int maxLength,
+                                    std::function<void(size_t length)> then,
+                                    int offset) {
+  if (maxLength <= 0) {
+    return then(offset);
+  }
+
+  // Get a string from the simulation memory and within the passed buffer
+  std::vector<char> data = memory_->getUntimedData(address + offset, maxLength);
+
+  for (int i = 0; i < data.size(); i++) {
+    buffer[i] = data[i];
+    // End of string; call onwards
+    if (buffer[i] == '\0') return then(i + 1);
+  }
+
+  // Reached max length; call onwards
+  return then(maxLength);
+}
+
+void SyscallHandler::readBufferThen(uint64_t ptr, uint64_t length,
+                                    std::function<void()> then) {
+  // If there's nothing to read, consider the read to be complete and call
+  // onwards
+  if (length == 0) {
+    return then();
+  }
+
```

Comment:
The `vAddrTranslation_` function should be invoked inside readBufferThen because it can return `faults::pagetable::IGNORED` or `faults::pagetable::DATA_ABORT`. Then return values should be handled accordingly.

Reply:
Moved translation to within function

Reply:
Given that we load the entire chunk, I think we should still do the translation outside the function but then skip the function call and return fatal if a data abort fault occurs. If an ignored region occurs then we can, in theory, return the correct sized read but with zero'ed out data. Made this change, please review.

Reply:
Happy with the changes, resolving this but left some extra comments